Plutarch/Field.hs

```diff
@@ -0,0 +1,190 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Field (
+  -- * PDataField class & deriving utils
+  PDataFields (..),
+  DerivePDataFields,
+  pletFields,
+  pletNFields,
+  pfield,
+
+  -- * BindFields class mechanism
+  BindFields (..),
+  type TermsOf,
+  type Take,
+
+  -- * Re-exports
+  HList (..),
+  HRec (..),
+  hlistField,
+  hrecField,
+) where
+
+import Data.Proxy (Proxy (..))
+import GHC.TypeLits (KnownNat, Nat, type (-))
+
+import Plutarch.Builtin (
+  PAsData,
+  PData,
+  PIsData (..),
+  pasConstr,
+  psndBuiltin,
+ )
+import Plutarch.DataRepr (
+  PDataRecord,
+  PIsDataRepr (..),
+  PLabeled (..),
+  pdhead,
+  pdtail,
+  pindexDataRecord,
+  type PNames,
+  type PTypes,
+ )
+import Plutarch.Field.HList (
+  HList (..),
+  HRec (..),
+  hlistField,
+  hrecField,
+  type IndexList,
+  type IndexOf,
+  type SingleItem,
+ )
+import Plutarch.Internal (TermCont (..), punsafeCoerce)
+import Plutarch.Prelude
+
+--------------------------------------------------------------------------------
+---------- PDataField class & deriving utils
+
+{- |
+  Class allowing 'letFields' to work for a PType, usually via
+  `PIsDataRepr`, but is derived for some other types for convenience.
+-}
+class PDataFields (a :: PType) where
+  -- | Fields in HRec bound by 'letFields'
+  type PFields a :: [PLabeled]
+
+  -- | Convert a Term to a 'PDataList'
+  ptoFields :: Term s a -> Term s (PDataRecord (PFields a))
+
+{- |
+  Derive PDataFields via a 'PIsDataRepr' instance,
+  using either 'Numbered' fields or a given list of fields.
+-}
+data DerivePDataFields (p :: PType) (s :: S)
+
+instance PDataFields (PDataRecord as) where
+  type PFields (PDataRecord as) = as
+  ptoFields = id
+
+instance
+  forall a as.
+  ( PIsDataRepr a
+  , SingleItem (PIsDataReprRepr a) ~ as
+  ) =>
+  PDataFields (DerivePDataFields a)
+  where
+  type
+    PFields (DerivePDataFields a) =
+      SingleItem (PIsDataReprRepr a)
+
+  ptoFields t =
+    (punsafeCoerce $ phoistAcyclic $ plam $ \d -> psndBuiltin #$ pasConstr # d)
+      # (punsafeCoerce t :: Term _ PData)
```

Comment:
not 100% on how sound this is, in theory if `PIsDataRepr` is being used properly then it should work this way but perhaps there's a safer way to impl this...

A `PlutusType` impl for `PIsDataReprInstances` would do the trick, I think.
Or maybe `(Coercible a (PDataRecord (PFields (DerivePDataFields a))))`